src/java/org/apache/cassandra/index/sai/disk/v1/sortedterms/TrieRangeIterator.java

```diff
@@ -0,0 +1,268 @@
+import org.apache.cassandra.io.tries.Walker;
+import org.apache.cassandra.io.util.Rebufferer;
+import org.apache.cassandra.utils.Pair;
+import org.apache.cassandra.utils.SizedInts;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+
+/**
+ * Thread-unsafe term -> long trie range iterator.
+ * The payload may be changed to something else.
+ */
+@NotThreadSafe
+class TrieRangeIterator extends Walker<TrieRangeIterator>
+{
+    private final ByteSource limit;
+    private final TrieTermsDictionaryReader.TransitionBytesCollector collector = new TrieTermsDictionaryReader.TransitionBytesCollector();
+    private final boolean inclEnd;
+    private IterationPosition stack;
+    private long next;
+    private boolean first = true;
+
+    public static class IterationPosition
+    {
+        public final long node;
+        public final int limit;
+        public final IterationPosition prev;
+
+        public int childIndex;
+
+        public IterationPosition(long node, int childIndex, int limit, IterationPosition prev)
+        {
+            super();
+            this.node = node;
+            this.childIndex = childIndex;
+            this.limit = limit;
+            this.prev = prev;
+        }
+
+        @Override
+        public String toString()
+        {
+            return String.format("[Node %d, child %d, limit %d]", node, childIndex, limit);
+        }
+    }
+
+    public TrieRangeIterator(Rebufferer source,
+                             long root,
+                             ByteComparable start,
+                             ByteComparable end,
+                             boolean admitPrefix,
+                             boolean inclEnd)
+    {
+        super(source, root);
+        limit = end != null ? end.asComparableBytes(BYTE_COMPARABLE_VERSION) : null;
+        this.inclEnd = inclEnd;
+
+        if (start != null)
+            initializeWithLeftBound(root, start.asComparableBytes(BYTE_COMPARABLE_VERSION), admitPrefix, limit != null);
+        else
+            initializeNoLeftBound(root, limit != null ? limit.next() : 256);
+    }
+
+    public Iterator<Pair<ByteSource, Long>> iterator()
+    {
+        return new AbstractIterator<Pair<ByteSource, Long>>()
+        {
+            @Override
+            protected Pair<ByteSource, Long> computeNext()
+            {
+                final long node = nextPayloadedNode();
+                if (node == -1 || getCurrentPayload() == -1)
+                {
+                    return endOfData();
+                }
+                ByteSource byteSource = collector.toByteComparable().asComparableBytes(ByteComparable.Version.OSS42);
+                return Pair.create(byteSource, getCurrentPayload());
+            }
+        };
+    }
+
+    public long getCurrentPayload()
+    {
+        return getPayload(buf, payloadPosition(), payloadFlags());
```

Comment:
nit: Could just inline this...

```
int bytes = payloadFlags();
return bytes == 0 ? TrieTermsDictionaryReader.NOT_FOUND : SizedInts.read(buf, payloadPosition(), bytes);
```
...and remove `getPayload()` altogether.

Reply:
As above